### ISSUE TITLE
approximate joint end-2-end training.

### ISSUE THREAD

**Issue by tornadomeet** [precedenceguo/mx-rcnn#23]
Hello, @precedenceguo 
Do you have a plan to add the approximate joint end-to-end training code?   i think we should first add the python op of propasal_target.py, which is something like ROIIter; then change a little of AnchorLoader.

**Comment by precedenceguo** [comment]
Not right now. The steps you suggest seems right. Have fun!

**Comment by tornadomeet** [comment]
ok, i will try this.

**Comment by neodooth** [comment]
@tornadomeet looking forward to your work. I took a try some days ago and found that the symbol variables are a terrible mess (rpn_bbox_target/bbox_target, etc).....

**Comment by tornadomeet** [comment]
@neodooth yes, for object detection, the io  part is the most tedious, because there is no generally io in mxnet for detection.

**Comment by precedenceguo** [comment]
@neodooth Another reason is that data parallelism requires such design to correctly determine correspondence for data, label and output.

**Comment by tornadomeet** [comment]
hello, @neodooth  i have written the preliminary end-to-end training code here: https://github.com/tornadomeet/mx-rcnn/blob/master/train_end2end.py but it will bring nan during training, the reason i think is that at the beginning training, the bbox_delta of rpn output  is explode, i will debug it more. glad to see you help for debug the end-to-end training. thanks~

**Comment by precedenceguo** [comment]
Please consider set `config.TRAIN.BBOX_NORMALIZATION_PRECOMPUTED = True` since Ross did that :) and indeed we should normalize that.

**Comment by tornadomeet** [comment]
@precedenceguo thanks, i had set it to be True yesterday, and nan still exists,  the output of rpn_bbox_pred will be nan sometimes, even when i set ``lr=0.0``` and ```wd=0.0```

**Comment by neodooth** [comment]
@tornadomeet Great!
bbox_delta does explodes, but if I set lr=0.0 it is OK, perhaps you hard coded some learning rate in the python scripts?

**Comment by tornadomeet** [comment]
@neodooth i found potential bugs in proposal_target.py/proposal.py, i am now debugging it.

**Comment by neodooth** [comment]
@tornadomeet I managed to draw a picture of the net, and found that you wrote two layers with the same name "rpn_cls_prob", one is SoftmaxOutput and the other is SoftmaxActivation. In rbg's implementation, he wrote 2 separate layers (Softmax and SoftmaxWithLoss

**Comment by tornadomeet** [comment]
@neodooth  i have fixed the name this morning, but not arrange the code so no push, i put the same the symbol, SoftmaxOutput and SoftmaxActivation

**Comment by tornadomeet** [comment]
@neodooth @precedenceguo  when training with joint end-to-end , at this line https://github.com/precedenceguo/mx-rcnn/blob/master/rcnn/rpn/proposal.py#L131, ```keep``` is like to be null, so how to fix this?

**Comment by neodooth** [comment]
@precedenceguo I changed RPN_MIN_SIZE to 5 in config.py. In my case, some imagenet images are extremely small, even smaller than the min size (16)

**Comment by tornadomeet** [comment]
@neodooth  ths, i have pushed the newest code i fix today, you can check that~

**Comment by neodooth** [comment]
I think the major problem lies in back-propagation and the losses, since if lr is set to 0 then everything is fine (except that the network is not learning). So the forwarding step seems good.

And I found that using a lr of 0.00001 is OK (resnet 101), I think this might be a clue. I remember in a paper it says that the network should be warmed up using a small lr.

**Comment by tornadomeet** [comment]
yes, i also think it needs warm up with a smaller lr.

**Comment by tornadomeet** [comment]
the NaN is also exists in py-faster-rcnn: https://github.com/rbgirshick/py-faster-rcnn/issues/65

**Comment by neodooth** [comment]
I added some BlockGrads to the net and tested 2 situations:
1. using rpn softmax to do the bp is ok
2. using rpn bbox smoothl1 to do the bp leads to NaN
This is strange since the difference between situation 2 and the original alternative rpn training is that there are more layers, while they don't contribute gradients

**Comment by tornadomeet** [comment]
@neodooth  i think there is no NaN during alternative rpn training is that it only use ```rpn_bbox_pred``` and ```bbox_target``` to bp in smooth_l1 loss, and not need to get the real box through bbox regression explicitly(this may leads to NaN).  
the reason of NaN i found is during the forward of proposal and proposal_target, the valid number of proposal box becomes zero, this maybe happen at any stage between their forward calc. 
i had add clip ```dw``` and ```dh``` and warm up lr_scheduler, but this helps a little.

**Comment by tornadomeet** [comment]
@neodooth  do you have a qq or email? if yes, we can discuss the problem there.

**Comment by neodooth** [comment]
@tornadomeet sent my qq to your email

**Comment by argman** [comment]
As i am not familar with mxnet-frcnn, but in py-faster-rcnn, NaN during training is because when you generate bbox targets, in pascal_voc.py line 208-211, you should pay attention to the -1.

**Comment by tornadomeet** [comment]
@argman, thanks, it is same in mx-rcnn: https://github.com/precedenceguo/mx-rcnn/blob/master/helper/dataset/pascal_voc.py#L126-L129

@precedenceguo @neodooth  i have solved the NaN problem, i 'll update it today.

**Comment by tornadomeet** [comment]
i have update my code, and the NaN will not come out during end2end training, i'll continue training a 2007 model to check the acc.

**Comment by abhiML** [comment]
Hey @tornadomeet could you give some pointers as to how you solved the problem? I am trying to solve a imilar problem in faster_rcnn_pytorch.
